### PR TITLE
Benchmark the preformance between Zig and C

### PR DESCRIPTION
**Pull request by samuel-schlaeppy**

### CODE REVIEW

File: docs/build-system.org [hunk 1]
```diff
@@ -189,3 +190,563 @@ All your codebase are belong to us.
 Run `zig build test` to run the tests.
 #+end_example
 
+
+** Performance comparaison & SMID
+
+The following chapters will explain different benchmark and code.
+
+*** Vectors
+A Zig vector is an array of child type: booleans, integers, floats or pointers.
```
Comment (Pismice) [inline docs/build-system.org]: it can only have those 4 types ?

File: docs/build-system.org [hunk 2]
```diff
@@ -189,3 +190,563 @@ All your codebase are belong to us.
...
+It provides SIMD instructions for parallelizing operations where possible.
+
+As described in the standard documentation, the vectors small than the target machine's SIMD instruction will be compile to a single SIMD instruction.
```
Comment (Pismice) [inline docs/build-system.org]: probably typo: smaller instead of small and "compiled" instead of "compile".

So that means that if the vectors are smaller there is not going to be optimized and treated as a "normal" array ?

File: docs/build-system.org [hunk 3]
```diff
@@ -189,3 +190,563 @@ All your codebase are belong to us.
...
+| c: 6                 |     8 |    10 |    12 |   |
+| d: 5                 |    12 |    21 |    32 |   |
+| e: 32                |   128 |   384 |  1024 |   |
+| f: false             | false | false | false |   |
+
+Zig support a vector size of:
+\begin{equation}
+2^{32} - 1
+\end{equation}
+But currently on the last Zig version, a long size (2^20) can crash at compile-time.
```
Comment (Pismice) [inline docs/build-system.org]: maybe add the link to this gh issue ?

File: docs/build-system.org [hunk 4]
```diff
@@ -189,3 +190,563 @@ All your codebase are belong to us.
...
+
+There are utility functions to simplify the use of vectors:
+1. ~@splat~, this function will create a vector of a given size with the same value:
+2. ~@Reduce~, this function will reduce the vector into a scalar. We can specify the operation.
```
Comment (Pismice) [inline docs/build-system.org:275]: maybe you could link the doc link to each of those functions

Comment (Pismice) [inline docs/build-system.org:275]: what is a scalar in zig ? is that a normal array ?

File: docs/build-system.org [hunk 5]
```diff
@@ -189,3 +190,563 @@ All your codebase are belong to us.
...
+   For intergers, every operation is available, for booleans plus there's ~.And~, ~.Or~, ~.Xor~.
+   And finally, for floating points they have also ~.Min~, ~.Max~, ~.Add~, ~.Mul~. [cite:@DocumentationZigProgramming]
+3. ~@shuffle~, this function will construct a new vector by selecting the elements of two vectors (a and b) with a mask.
```
Comment (Pismice) [inline docs/build-system.org]: what is a mask ?

File: docs/build-system.org [hunk 6]
```diff
@@ -189,3 +190,563 @@ All your codebase are belong to us.
...
+   If the number of the mask is 0 or upper, it will select the index of the first vector (a).
+   If the number of the mask is -1 or less, it will select the index starting at -1 and decrementing. (read example to be more clear)
+4. ~@select~, this function will select from two vectors (a and b) based on a ~pred~ vector.
+   If ~pred[i]~ is positif, the element will be ~a[i]~ otherwise ~b[i]~
+   
+#+begin_src zig :imports '(std) :main no
+  const stdout = std.io.getStdOut();
```
Comment (Pismice) [inline docs/build-system.org:286]: why doing that instead of debug.print ?

File: docs/build-system.org [hunk 7]
```diff
@@ -189,3 +190,563 @@ All your codebase are belong to us.
...
+\begin{equation}
+\frac{\pi}{4} = 1 - \frac{1}{3} + \frac{1}{5} - \frac{1}{7} + \frac{1}{9} - \ldots = \sum_{k = 0}^{\infty} \frac{(-1)^k}{2k + 1}
+\end{equation}
+
+The C code shows the standard way to compute the Leibniz algorithm, at each iteration, PI will improve.
+#+begin_src C :includes <stdio.h> :tangle leibniz.c
+  unsigned rounds = 1000000000;
+  double pi = 1.0;
+
+  rounds += 2u; // do this outside the loop
```
Comment (Pismice) [inline docs/build-system.org:349]: why doing this outside the loop ?

Comment (Pismice) [reply docs/build-system.org:349]: why not directly having 1000000002

File: docs/build-system.org [hunk 8]
```diff
@@ -189,3 +190,563 @@ All your codebase are belong to us.
...
+
+  for (unsigned i=2u; i < rounds; ++i) // use ++i instead of i++
```
Comment (Pismice) [inline docs/build-system.org]: why adding u ?

Comment (Pismice) [reply docs/build-system.org]: and will ++i really change anything?

File: docs/build-system.org [hunk 9]
```diff
@@ -189,3 +190,563 @@ All your codebase are belong to us.
...
+    {
+      double x = -1.0 + 2.0 * (i & 0x1); // allows vectorization
```
Comment (Pismice) [inline docs/build-system.org:353]: why does it allow vectorization ?

File: docs/build-system.org [hunk 10]
```diff
@@ -189,3 +190,563 @@ All your codebase are belong to us.
...
+  pi *= 4;
+  try std.io.getStdOut().writer().print("{}", .{pi});
+#+end_src
+
+#+RESULTS:
+: 3.1415926545880506
+
+This Zig optmized code will produce a simalar output that the two other codes but it use SIMD instruction to try to manually optimize the code.
+Now the number of iteration is dived by 4, because the calculations are parallelized by 4.
+Note: you can see that the length of a ~vector~ is 4.
```
Comment (Pismice) [inline docs/build-system.org]: how can we see that ?

File: docs/build-system.org [hunk 11]
```diff
@@ -189,3 +190,563 @@ All your codebase are belong to us.
...
+  try std.io.getStdOut().writer().print("{}", .{pi});
+#+end_src
+
+#+RESULTS:
+: 3.1415926545880506
+
+This Zig optmized code will produce a simalar output that the two other codes but it use SIMD instruction to try to manually optimize the code.
+Now the number of iteration is dived by 4, because the calculations are parallelized by 4.
+Note: you can see that the length of a ~vector~ is 4. 
+And the end of the code, we find a code that is not vectorized, it will only use for the last iterations.
```
Comment (Pismice) [inline docs/build-system.org]: not sure to understand this sentence

File: docs/build-system.org [hunk 12]
```diff
@@ -189,3 +190,563 @@ All your codebase are belong to us.
...
+#+begin_example
+Benchmark 1: ./leibniz_c
+  Time (mean ± σ):     458.2 ms ±   3.9 ms    [User: 457.3 ms, System: 0.3 ms]
+  Range (min … max):   452.8 ms … 465.4 ms    10 runs
+ 
+Benchmark 2: ./leibniz_zig
+  Time (mean ± σ):     898.0 ms ±   1.6 ms    [User: 896.9 ms, System: 0.0 ms]
+  Range (min … max):   894.1 ms … 899.7 ms    10 runs
+ 
+Benchmark 3: ./leibniz_simd_zig
```
Comment (Pismice) [inline docs/build-system.org:464]: why is it taking the same time as leibniz_c that is not using simd ?

File: docs/build-system.org [hunk 13]
```diff
@@ -189,3 +190,563 @@ All your codebase are belong to us.
...
+#+end_example
+
+If you run the benchmark in your computer, you can see the result in the block upper.
+
+*** ARM benchmark example
+
+We made the same benchmark but for an arm CPU, here are the compilers's options used.
+The target is a nanopi neo plus 2.
+#+begin_src shell
+  aarch64-buildroot-linux-gnu-gcc leibniz.c -o leibniz_c -O3 -march=armv8-a -mtune=cortex-a53 -s -fno-signed-zeros -fno-trapping-math -fassociative-math
```
Comment (Pismice) [inline docs/build-system.org:481]: why did you add all those specific build options ?

File: docs/build-system.org [hunk 14]
```diff
@@ -189,3 +190,563 @@ All your codebase are belong to us.
...
+  zig build-exe leibniz_smid.zig -OReleaseFast -femit-bin=leibniz_zig_simd_aarch -target aarch64-native -mcpu cortex_a53 -fstrip
+  zig build-exe leibniz.zig -OReleaseFast -femit-bin=leibniz_zig_aarch -target aarch64-native -mcpu cortex_a53 -fstrip
+#+end_src
+
+#+RESULTS:
+
+On our arm target we cannot execute ~hyperfine~, we measure 8 times the each program and we made statistics.
```
Comment (Pismice) [inline docs/build-system.org:488]: why is it so different than on x86 ?

File: docs/build-system.org [hunk 15]
```diff
@@ -189,3 +190,563 @@ All your codebase are belong to us.
...
+|----------------------------+----------+---------+---------|
+| =./leibniz_c=              |    20.57 |   20.58 |   20.57 |
+| =./leibniz_zig_aarch=      |    19.95 |   19.96 |   19.95 |
+| =./leibniz_zig_simd_aarch= |    14.43 |   14.44 |   14.43 |
+
+
+*** TODO Differents code examples benchmark
+TODO: update the benchmarks with the compiler commands
+
+Mes benchmarks sur mon ordi :
```
Comment (Pismice) [inline docs/build-system.org:505]: Should be in english ? Or is it a draft ?

File: docs/build-system.org [hunk 16]
```diff
@@ -189,3 +190,563 @@ All your codebase are belong to us.
...
+Their website shows benchmarks made with their hardware [[https://programming-language-benchmarks.vercel.app/zig-vs-c][zig-vs-c]]. [cite:@ZigVSBenchmarks]
+We then reproduced the benchmarks between the C and the Zig to see if there were any differences with our hardware.
+
+Here's our results:
+CPU INFO:[x86\_64][8 cores] Intel(R) Core(TM) i7-8565U CPU @ 1.80GHz (Model 142)
+
+Here's what you need to know about file naming:
+#+begin_quote
+- *-m in a file name stands for multi-threading or multi-processing
+- *-i in a file name stands for direct intrinsics usage. (Usage of simd intrinsics via libraries is not counted)
```
Comment (Pismice) [inline docs/build-system.org]: what is direct intrinsics usage ? is the read supposed to know ?

Comment (Pismice) [reply docs/build-system.org]: same question for FFI

File: docs/build-system.org [hunk 17]
```diff
@@ -189,3 +190,563 @@ All your codebase are belong to us.
...
+| c    | [[https://github.com/hanabi1224/Programming-Language-Benchmarks/blob/main/bench/algorithm/spectral-norm/3.c][3-m.c]]   | 78ms  | 10ms   | 2.1MB        | 600ms      | 0ms       | gcc 13.2.0                |
+| c    | [[https://github.com/hanabi1224/Programming-Language-Benchmarks/blob/main/bench/algorithm/spectral-norm/4.c][4-m.c]]   | 80ms  | 12ms   | 2.1MB        | 590ms      | 0ms       | gcc 13.2.0                |
+| c    | [[https://github.com/hanabi1224/Programming-Language-Benchmarks/blob/main/bench/algorithm/spectral-norm/5-im.c][5-im.c]]  | 81ms  | 13ms   | 2.3MB        | 620ms      | 0ms       | gcc 13.2.0                |
+| zig  | [[https://github.com/hanabi1224/Programming-Language-Benchmarks/blob/main/bench/algorithm/spectral-norm/1.zig][1.zig]]   | 192ms | 3.6ms  | 1.3MB        | 203ms      | 0ms       | zig 0.12.0                |
+
+
+*** Results
+We can see that execution times vary greatly depending on the programming language, the implementation, and finally, the compiler used.
```
Comment (Pismice) [inline docs/build-system.org]: and the target ?

File: docs/build-system.org [hunk 18]
```diff
@@ -189,3 +190,563 @@ All your codebase are belong to us.
...
+In addition, we found that the best-performing Zig codes use vectors that are optimised by the compiler.
```
Comment (Pismice) [inline docs/build-system.org]: what do you mean by "optimised by the compiler", are there vectors that are optimised and other not optimised ?

File: docs/build-system.org [hunk 19]
```diff
@@ -189,3 +190,563 @@ All your codebase are belong to us.
...
+On the other hand, Zig codes that are not optimised tend to perform less well than C codes.
+
+On the other hand, we can see that Zig codes consume less memory than C codes.
```
Comment (Pismice) [inline docs/build-system.org:651]: why ?

File: docs/build-system.org [hunk 20]
```diff
@@ -189,3 +190,563 @@ All your codebase are belong to us.
...
+
+
+*** Results
+We can see that execution times vary greatly depending on the programming language, the implementation, and finally, the compiler used.
+In addition, we found that the best-performing Zig codes use vectors that are optimised by the compiler.
+On the other hand, Zig codes that are not optimised tend to perform less well than C codes.
+
+On the other hand, we can see that Zig codes consume less memory than C codes. 
+But for the comparison to be fair, the allocator used in Zig is the libC allocator.
+And finally, we can see that the best-performing code depend also on the input.
```
Comment (Pismice) [inline docs/build-system.org]: maybe summarize, what type of inputs work better than others

File: docs/build-system.org [hunk 21]
```diff
@@ -189,3 +190,563 @@ All your codebase are belong to us.
...
+| zig  | [[https://github.com/hanabi1224/Programming-Language-Benchmarks/blob/main/bench/algorithm/spectral-norm/1.zig][1.zig]]   | 192ms | 3.6ms  | 1.3MB        | 203ms      | 0ms       | zig 0.12.0                |
+
+
+*** Results
+We can see that execution times vary greatly depending on the programming language, the implementation, and finally, the compiler used.
+In addition, we found that the best-performing Zig codes use vectors that are optimised by the compiler.
+On the other hand, Zig codes that are not optimised tend to perform less well than C codes.
+
+On the other hand, we can see that Zig codes consume less memory than C codes. 
+But for the comparison to be fair, the allocator used in Zig is the libC allocator.
```
Comment (Pismice) [inline docs/build-system.org]: you added "But", but is it really related to the last sentence ?

File: docs/build-system.org [hunk 22]
```diff
@@ -189,3 +190,563 @@ All your codebase are belong to us.
...
+  4. zigcc 0.12.0-dev.2341+92211135f
+
+For the knucleotide example, we obtain a comparable ranking in the two results, however, we notice that my x86 target is 85 ms slower for the version of Zig with an input of 2500000, but it consumes less memory (around 500KB).
+
+For the nbody example, we can note that on average, the performance obtained is better on my x86 target, but the memory used remains in the same order of magnitude.
+We also see that the ~8-i.c~ implementation is faster under AMD Epics for the ~zigcc~ compiler and for ~gcc~.
+And that the ~5.c~ implementation with the ~Clang~ compiler is among the first in the x86 test, but is among the last for AMD.
+We note that the difference between version 14 and version 17 of clang can also affect the results.
+This is the only example where the zigcc compiler is used for benchmarking, and we can see that the performance results obtained are similar to those we can get with gcc or clang.
+We can also observe that the same code compiled with zigcc uses less memory than the other compilers.
```
Comment (Pismice) [inline docs/build-system.org:677]: why ? is there a reason that can be mentionned ?

File: docs/build-system.org [hunk 23]
```diff
@@ -189,3 +190,563 @@ All your codebase are belong to us.
...
+
+After each implementation build, we benchmark with hyperfine to have the results:
+
+1. First, we used =hyperfine= to benchmark the three implementations one after the other:
+#+begin_src shell
+  hyperfine --warmup=3 './1brc/bin/analyze ~/Documents/1brc/measurements.txt' './1brc-zig/zig-out/bin/1brc-zig ~/Documents/1brc/measurements.txt' './1brc-zig2/zig-out/bin/1brc-zig ~/Documents/1brc/measurements.txt' --export-orgmode "benchmark-1brc.org" --output null -i
+#+end_src
+
+And we have the results:
+| Command                                                              | Mean [s]       | Min [s] | Max [s] | Relative    |
```
Comment (Pismice) [inline docs/build-system.org:706]: instead of writing the full command it could be more readable to write something like:
- C implementation
- 1th zig implementation
- 2th zig implemnation

File: docs/build-system.org [hunk 24]
```diff
@@ -189,3 +190,563 @@ All your codebase are belong to us.
...
+|----------------------------------------------------------------------+----------------+---------+---------+-------------|
+| =./1brc/bin/analyze ~/Documents/1brc/measurements.txt=               | 11.774 ± 2.051 |   7.582 |  13.034 | 1.00        |
+| =./1brc-zig/zig-out/bin/1brc-zig ~/Documents/1brc/measurements.txt=  | 21.283 ± 0.051 |  21.228 |  21.390 | 1.81 ± 0.31 |
+| =./1brc-zig2/zig-out/bin/1brc-zig ~/Documents/1brc/measurements.txt= | 24.788 ± 0.187 |  24.332 |  24.931 | 2.11 ± 0.37 |
+
+We can see the differences in performances between C and Zig, but we note that the differences between the best C result and the worst C result have a difference of 6 seconds.
```
Comment (Pismice) [inline docs/build-system.org]: I dont think the usage of "but" is good, because your second part is not related to the first. "but" should be used to mark a contradiction

File: docs/build-system.org [hunk 25]
```diff
@@ -189,3 +190,563 @@ All your codebase are belong to us.
...
+
+To see if we get different performances, we benchmark each implementation on its own:
+
+The first Zig implementation [[https://github.com/candrewlee14/1brc-zig][1brc-zig]]
+| Command                                        | Mean [s]       | Min [s] | Max [s] | Relative |
+|------------------------------------------------+----------------+---------+---------+----------|
```
Comment (Pismice) [inline docs/build-system.org]: why does it produce different results to the ones you did 10 lines on top 

isnt that the same: ./1brc-zig/zig-out/bin/1brc-zig ~/Documents/1brc/measurements.txt ?

File: docs/build-system.org [hunk 26]
```diff
@@ -189,3 +190,563 @@ All your codebase are belong to us.
...
+| Command                                       | Mean [s]      | Min [s] | Max [s] | Relative |
+|-----------------------------------------------+---------------+---------+---------+----------|
+| =./analyze ~/Documents/1brc/measurements.txt= | 8.898 ± 0.591 |   7.507 |   9.510 |     1.00 |
+
+
+In the C implementation, we note that his writer uses a custom hashmap to hash the city names.
+He uses a fast multiplication hash combined with a linear search probing.
+The load factor of his hashmap is 0.5. [cite:@BillionRowChallenge]
+
+We find out that in the first Zig implementation, use also a hashmap but the Zig standard library hashmap for string hash with the wyhash [cite:@wangyiWangyifudanWyhash2024] algorithm.
```
Comment (Pismice) [inline docs/build-system.org]: you might have to rephrase

File: docs/build-system.org [hunk 27]
```diff
@@ -189,3 +190,563 @@ All your codebase are belong to us.
...
+[[../images/flamegrah_1brc_c.png]]
+
+In the example below, this is how the hash is calculated for each city: [cite:@OneBillionRows2024a]
+#+begin_example c
+// hash everything up to ';'
+// assumption: key (city name) is at least 1 char
+unsigned int len = 1;
+unsigned int hash = (unsigned char)buf[0];
+while (buf[len] != ';') {
+  hash = (hash * 31) + (unsigned char)buf[len++];
```
Comment (Pismice) [inline docs/build-system.org:774]: is that normal that i dont really understand what is going on here ? oupsi

File: docs/build-system.org [hunk 28]
```diff
@@ -189,3 +190,563 @@ All your codebase are belong to us.
...
+
+// idx is now either 0 (new entry)
+// or contains the index of our key in the results a
+#+end_example
+
+
+For the second implementation of the Zig code, we note that it has been implemented with a =btree=, used for data merge.
+This merge is performed because there are several threads processing the data.
+However, it also uses the same hashmap as the one used in the first version.
+
```
Comment (Pismice) [inline docs/build-system.org:792]: Maybe you should add a conclusion to wrap it all ?

**Review by Pismice — CHANGES_REQUESTED** [review]

File: docs/build-system.org (hunk 1 above)
Comment (samuel-schlaeppy) [reply docs/build-system.org]: Yes: https://ziglang.org/documentation/0.12.0/#Vectors

File: docs/build-system.org (hunk 4 above)
Comment (samuel-schlaeppy) [reply docs/build-system.org:275]: Scalar in math, like a unique value.

File: docs/build-system.org (hunk 5 above)
Comment (samuel-schlaeppy) [reply docs/build-system.org]: It's a array that will select the good array and index to select. Maybe I can say that's also an array

File: docs/build-system.org (hunk 6 above)
Comment (samuel-schlaeppy) [reply docs/build-system.org:286]: It will not print the result on the org file

File: docs/build-system.org (hunk 7 above)
Comment (samuel-schlaeppy) [reply docs/build-system.org:349]: I didn't write the C code, but I adapted it to work in org. Originally, the choice of iteration is made via the program arguments

File: docs/build-system.org (hunk 8 above)
Comment (samuel-schlaeppy) [reply docs/build-system.org]: Like the last comment, It's not the code I wrote, but the `2u` is to say that it's a 2 unsigned

File: docs/build-system.org (hunk 9 above)
Comment (samuel-schlaeppy) [reply docs/build-system.org:353]: It will help to use the asm instruction `fuseAddMultiply`. Maybe I can explain why the wrote that

File: docs/build-system.org (hunk 10 above)
Comment (samuel-schlaeppy) [reply docs/build-system.org]: In the vector, it's `@vector(size, type)` in the example below

File: docs/build-system.org (hunk 13 above)
Comment (samuel-schlaeppy) [reply docs/build-system.org:481]: It's to cross-compile to the nanopi

Comment (samuel-schlaeppy) [reply docs/build-system.org:481]: Specify the target architecture and use the fast math

File: docs/build-system.org (hunk 14 above)
Comment (samuel-schlaeppy) [reply docs/build-system.org:488]: The speed difference ? Because it's not the same set of instruction

File: docs/build-system.org (hunk 15 above)
Comment (samuel-schlaeppy) [reply docs/build-system.org:505]: Yes it's a draft

File: docs/build-system.org (hunk 19 above)
Comment (samuel-schlaeppy) [reply docs/build-system.org:651]: It will be difficult to explain why there is a difference in memory, it can depend of compiler version, Glibc version or implementation, the use of functions that allocate more memory, ...

File: docs/build-system.org (hunk 20 above)
Comment (samuel-schlaeppy) [reply docs/build-system.org]: It's difficult because it change on the problem, the compiler, the target, and the size of the input

Comment (samuel-schlaeppy) [reply docs/build-system.org]: Maybe I haven't understood your request

File: docs/build-system.org (hunk 22 above)
Comment (samuel-schlaeppy) [reply docs/build-system.org:677]: It's an observation, I cannot generalize this information, but it's interesting to know. (The last sentence will clarify the observation)

File: docs/build-system.org (hunk 3 above)
Comment (Pismice) [reply docs/build-system.org]: is it done ? if yes you can just react to my message with an emoji

File: docs/build-system.org (hunk 5 above)
Comment (Pismice) [reply docs/build-system.org]: how does the mask decide which is the good array ?

File: docs/build-system.org (hunk 7 above)
Comment (Pismice) [reply docs/build-system.org:349]: maybe you could rewrite it so that readers wont be confused the same way I was

File: docs/build-system.org (hunk 8 above)
Comment (Pismice) [reply docs/build-system.org]: mhhhh ok I think you should tweak the exemple of bit to make it a bit more coherent, and remove useless comments like "// use ++i instead of i++" because its useless

File: docs/build-system.org (hunk 9 above)
Comment (Pismice) [reply docs/build-system.org:353]: is it somehow related to AVX ?

File: docs/build-system.org (hunk 10 above)
Comment (Pismice) [reply docs/build-system.org]: so you should not say "a" vector but "the" vector ?

File: docs/build-system.org (hunk 19 above)
Comment (Pismice) [reply docs/build-system.org:651]: maybe its could be useful to precise that and add something like: "i dont know for sure but it could be linked to blablabla...." ?

File: docs/build-system.org (hunk 20 above)
Comment (Pismice) [reply docs/build-system.org]: from what I understood you implied that the best-performing code can be different depending on the input, is that right ? if yes you should say what type of input will prefer which implementation

File: docs/build-system.org (hunk 5 above)
Comment (samuel-schlaeppy) [reply docs/build-system.org]: It's explain in below. It's not clear ?

File: docs/build-system.org (hunk 9 above)
Comment (samuel-schlaeppy) [reply docs/build-system.org:353]: Yes, AVX it a type of instruction set

File: docs/build-system.org (hunk 11 above)
Comment (samuel-schlaeppy) [reply docs/build-system.org]: Now it's better ?

File: docs/build-system.org (hunk 12 above)
Comment (samuel-schlaeppy) [reply docs/build-system.org:464]: Now it's better ? (the explanation is upper in the doc)

File: docs/build-system.org (hunk 16 above)
Comment (samuel-schlaeppy) [reply docs/build-system.org]: I explain for intrinsics and I remove the FFI, it's used only in the benchmark site

File: docs/build-system.org (hunk 18 above)
Comment (samuel-schlaeppy) [reply docs/build-system.org]: Now it's better ?

File: docs/build-system.org (hunk 20 above)
Comment (samuel-schlaeppy) [reply docs/build-system.org]: In fact the input is the number of iteration of the alogrithm

File: docs/build-system.org (hunk 5 above)
Comment (Pismice) [reply docs/build-system.org]: yes sorry indeed very clear, I just hadnt those lines in the review UI :)

File: docs/build-system.org (hunk 26 above)
Comment (samuel-schlaeppy) [reply docs/build-system.org]: Now it's better ?

File: docs/build-system.org (hunk 27 above)
Comment (samuel-schlaeppy) [reply docs/build-system.org:774]: Upper, I explain more in depth the example. It's better ?

File: docs/build-system.org (hunk 11 above)
Comment (Pismice) [reply docs/build-system.org]: love it

File: docs/build-system.org (hunk 12 above)
Comment (Pismice) [reply docs/build-system.org:464]: Could not find it :/ would you mind giving me the link ok line number

File: docs/build-system.org (hunk 14 above)
Comment (Pismice) [reply docs/build-system.org:488]: yes but arm is 20sec and x86 is like 0.5sec. is x86 just more performant ?

File: docs/build-system.org (hunk 12 above)
Comment (samuel-schlaeppy) [reply docs/build-system.org:464]: Line 328 of the `build-system.org`

File: docs/build-system.org (hunk 14 above)
Comment (samuel-schlaeppy) [reply docs/build-system.org:488]: It's not the same target, the x86 is my laptop and the arm is a nanopi -> [CPU of the nanopi](https://mse-csel.github.io/website/documentation/assets/DDI0500D_cortex_a53_r0p2_trm.pdf)

File: docs/build-system.org (hunk 24 above)
Comment (samuel-schlaeppy) [reply docs/build-system.org]: It's better now ?

File: docs/build-system.org (hunk 28 above)
Comment (samuel-schlaeppy) [reply docs/build-system.org:792]: Yes, maybe I will add it in another PR

File: docs/build-system.org (hunk 19 above)
Comment (samuel-schlaeppy) [reply docs/build-system.org:651]: Yes but I don't have a good idea what cause the difference. I don't want to tell some nonsense

File: docs/build-system.org (hunk 7 above)
Comment (samuel-schlaeppy) [reply docs/build-system.org:349]: It's because PI is initialize at 1 : `pi = 1`.
And to calculate PI you need:

$$ \frac{\pi}{4} = 1 - \frac{1}{3} + \frac{1}{5} - \frac{1}{7} + \frac{1}{9} - \ldots$$

And if `i` start at `0` the calculation will not work.
That's why we add `2`.

Comment (Pismice) [reply docs/build-system.org:349]: yes but you added 2 to `rounds` not `pi`

File: docs/build-system.org (hunk 19 above)
Comment (Pismice) [reply docs/build-system.org:651]: you could rewrite this "In addition, we found that the best-performing Zig codes use vectors." to something like "In additin, we found that the best-performing code is the Zig code using vectors" ?

File: docs/build-system.org (hunk 20 above)
Comment (Pismice) [reply docs/build-system.org]: yes, but with your sentence you imply that your "best-performing code" (in our case the zig with vectors) isnt always the best, but it always is from what i see from the benchmarks no ?

File: docs/build-system.org (hunk 7 above)
Comment (samuel-schlaeppy) [reply docs/build-system.org:349]: Yes because in the loop, we start at 2 until rounds, and to have to good amount of rounds, we add 2.
```c
for (unsigned i=2u; i < rounds; ++i)
```

File: docs/build-system.org (hunk 19 above)
Comment (samuel-schlaeppy) [reply docs/build-system.org:651]: I change the line 635, it's better ?

File: docs/build-system.org (hunk 20 above)
Comment (samuel-schlaeppy) [reply docs/build-system.org]: I change it to say about the performances. Check the new version

**Review by Pismice — APPROVED** [review]
